servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectChannelSingle.java

```diff
@@ -125,26 +128,35 @@ private void sendConnectRequest(final ChannelHandlerContext ctx) {
         public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
             if (msg instanceof HttpResponseMetaData) {
                 if (response != null) {
-                    failSubscriber(ctx, new RetryableProxyConnectException(
-                            "Received two responses for a single CONNECT request"));
+                    failSubscriber(ctx, new RetryableProxyConnectException(ctx.channel() +
+                            " Received two responses for a single CONNECT request"));
                     return;
                 }
                 response = (HttpResponseMetaData) msg;
                 if (response.status().statusClass() != SUCCESSFUL_2XX) {
-                    failSubscriber(ctx, new ProxyResponseException("Non-successful response '" + response.status() +
-                            "' from proxy on CONNECT " + connectAddress, response.status()));
+                    failSubscriber(ctx, unsuccessfulResponse(ctx.channel(), response, connectAddress));
                 }
                 // We do not complete subscriber here because we need to wait for the HttpResponseDecoder state machine
                 // to complete. Completion will be signalled by InboundDataEndEvent. Any other messages before that are
                 // unexpected, see https://datatracker.ietf.org/doc/html/rfc9110#section-9.3.6
                 // It also helps to make sure we do not propagate InboundDataEndEvent after the next handlers are added
                 // to the pipeline, potentially causing changes in their state machine.
             } else {
-                failSubscriber(ctx, new RetryableProxyConnectException(
-                        "Received unexpected message in the pipeline of type: " + msg.getClass().getName()));
+                failSubscriber(ctx, new RetryableProxyConnectException(ctx.channel() +
+                        " Received unexpected message in the pipeline of type: " + msg.getClass().getName()));
             }
         }
 
+        private static ProxyConnectResponseException unsuccessfulResponse(final Channel channel,
+                                                                          final HttpResponseMetaData response,
+                                                                          final String connectAddress) {
+            final String message = channel + " Non-successful response '" + response.status() +
+                    "' from proxy on CONNECT " + connectAddress;
+            return response.status().statusClass() == SERVER_ERROR_5XX ?
```

Comment:
I wonder if this is too granular 🤔 .. i.e. retrying a 501 is likely not helping. We could go more conservative and choose a list (i.e. 500, 502 503, 504), or ultimately just do not mark anything as retryable and let the user decide?

Reply:
good point! adjusted